File: Prid1920-g03/Models/Post.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Diagnostics;
using Microsoft.EntityFrameworkCore;

namespace Prid1920_g03.Models
{

    public class Post
    {
        [Key]
        public int Id { get; set; }
        public string Title { get; set; }
        public string Body { get; set; }
        public DateTime Timestamp { get; set; }
               
        public int? ParentId { get; set; }
        public int AuthorId { get; set; } 
        public int? AcceptedAnswerId { get; set; }
        
        public virtual Post PostParent { get; set; }
        public virtual Post AcceptedAnswer { get; set; }
        public virtual User User { get; set; }
        public virtual IList<Post> Responses { get; set; } = new List<Post>();
        public virtual IList<Comment> Comments { get; set; } = new List<Comment>();
        public virtual IList<Vote> Votes { get; set; } = new List<Vote>();

        public virtual IList<PostTag> PostTags { get; set; } = new List<PostTag>();

        [NotMapped]
        public IEnumerable<Tag> Tags { get => PostTags.Select(t => t.Tag); }

        [NotMapped]
        public int NumResponse
        {
            get
            {
                return(from c in Responses select c).Count();
            }
        }

        [NotMapped]
        public int VoteState
        {
            get
            {
                return Votes.Sum(r => r.UpDown);
            }
        }

        [NotMapped]
        public int HightVote
        {
            get
            {
                return Math.Max(VoteState, Responses.Count > 0 ? Responses.Max(r => r.VoteState) : VoteState);
            }        
        }

        [NotMapped]
        public int NumVote
        {
            get
            {
                return(from v in Votes select v ).Count();
            }
        }

        [NotMapped]
        public int NumComment
        {
            get
            {
                return(from c in Comments select c).Count();
            }
        }
    }
}

File: Prid1920-g03/ClientApp/src/app/models/post.ts
import { Vote } from "./vote";

  export class Post {
      id: number;
      title: string;
      body: string;
      timestamp: string;
      authorId: number;
      parentId: number;
      acceptedAnswerId: number;
      acceptedAnswerIdExist: boolean;

      user: any;
      postParent: any;
      numResponse: number;
      numVote: number;
      numComments: number;
      voteState: number;
      hightVote: number;
      
      comments: any[];
      responses: (string | Post)[];
      tags: string[];
      votes: Vote[];
    
      constructor(data: any) {
        if (data) {
          this.id = data.id;
          this.title = data.title;
          this.body = data.body;
          this.timestamp = data.timestamp;
          
          this.authorId = data.authorId;
          this.parentId = data.parentId;
          this.acceptedAnswerId = data.acceptedAnswerId;          
          
          this.user = data.user;
          this.numResponse = data.numResponse,
          this.hightVote = data.hightVote,
          this.voteState = data.voteState,
          this.numVote = data.numVote,
          this.numComments = data.numComments,

          this.comments = data.comments;
          this.responses = data.replies;
          this.tags = data.tags;
          this.votes = data.votes;
        }
      }
    }  
  
  